hail/src/main/scala/is/hail/services/batch_client/BatchClient.scala

```diff
@@ -0,0 +1,163 @@
+import org.apache.http.client.methods.{HttpDelete, HttpGet, HttpPatch, HttpPost, HttpUriRequest}
+import org.apache.http.entity.{ByteArrayEntity, ContentType, StringEntity}
+import org.apache.http.impl.client.{CloseableHttpClient, HttpClients}
+import org.apache.http.util.EntityUtils
+import org.json4s.JsonAST.JNull
+import org.json4s.{DefaultFormats, Formats, JObject, JValue}
+import org.json4s.jackson.JsonMethods
+
+class NoBodyException(message: String, cause: Throwable) extends Exception(message, cause) {
+  def this() = this(null, null)
+
+  def this(message: String) = this(message, null)
+}
+
+class ClientResponseException(statusCode: Int, message: String, cause: Throwable) extends Exception(message, cause) {
+  def this(statusCode: Int) = this(statusCode, null, null)
+
+  def this(statusCode: Int, message: String) = this(statusCode, message, null)
+}
+
+class BatchClient extends AutoCloseable {
+  private[this] val baseUrl = DeployConfig.get.baseUrl("batch")
+
+  private[this] val httpClient: CloseableHttpClient = HttpClients.createDefault();
+
+  private[this] def request(req: HttpUriRequest, body: HttpEntity = null): JValue = {
+    if (body != null)
+      req.asInstanceOf[HttpEntityEnclosingRequest].setEntity(body)
+    Tokens.get.addServiceAuthHeaders("batch", req)
+    using(httpClient.execute(req)) { resp =>
+      val statusCode = resp.getStatusLine.getStatusCode
+      if (statusCode < 200 || statusCode >= 300) {
+        val entity = resp.getEntity
+        val message =
+          if (entity != null)
+            EntityUtils.toString(entity)
+        else
+            null
+        throw new ClientResponseException(statusCode, message)
+      }
+      val entity: HttpEntity = resp.getEntity
+      if (entity != null) {
+        using(entity.getContent) { content =>
+          val s = IOUtils.toByteArray(content)
+          if (s.isEmpty)
+            null
+          else
+            JsonMethods.parse(new String(s))
+        }
+      } else
+        null
+    }
+  }
+
+  def get(path: String): JValue =
+    request(new HttpGet(s"$baseUrl$path"))
+
+  def post(path: String, body: HttpEntity): JValue =
+    request(new HttpPost(s"$baseUrl$path"), body = body)
+
+  def post(path: String, json: JValue = null): JValue =
+    post(path,
+      if (json != null)
+        new StringEntity(
+          JsonMethods.compact(json),
+          ContentType.create("application/json"))
+      else
+        null)
+
+  def patch(path: String): JValue =
+    request(new HttpPatch(s"$baseUrl$path"))
+
+  def delete(path: String): JValue =
+    request(new HttpDelete(s"$baseUrl$path"))
+
+  def createJobs(batchID: Long, jobs: IndexedSeq[JObject]): Unit = {
+    val bunches = new ArrayBuilder[Array[Array[Byte]]]()
+
+    val bunchb = new ArrayBuilder[Array[Byte]]()
+
+    var i = 0
+    var size = 0
+    while (i < jobs.length) {
+      val jobBytes = JsonMethods.compact(jobs(i)).getBytes
```

Comment:
This is maybe pedantic but can we specify the UTF-8 charset? Python [unambiguously defaults to utf-8](https://docs.python.org/3.7/library/stdtypes.html#bytes.decode) whereas Java claims the charset depends on the locale of the VM.

Reply:
No, I agree, I think we should use UTF-8 everywhere.